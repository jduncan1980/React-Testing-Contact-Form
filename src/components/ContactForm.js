import React, { useState } from 'react';
import { useForm } from 'react-hook-form';

const ContactForm = () => {
	const [data, setData] = useState();

	const { register, errors, handleSubmit, formState } = useForm({
		mode: 'onChange',
	});
	const onSubmit = (data) => {
		setData(data);
	};

	return (
		<div className='App'>
			<form onSubmit={handleSubmit(onSubmit)} data-testid='form'>
				<div>
					<label htmlFor='firstName' data-testid='firstNameLabel'>
						First Name*
					</label>
					<input
						data-testid='firstNameInput'
						name='firstName'
						placeholder='First Name'
						ref={register({ required: true })}
					/>
					{errors.firstName && (
						<p>Looks like there was an error: {errors.firstName.type}</p>
					)}
				</div>

				<div>
					<label htmlFor='lastName' data-testid='lastNameLabel'>
						Last Name*
					</label>
					<input
						data-testid='lastNameInput'
						name='lastName'
						placeholder='Last Name'
						ref={register({ required: true })}
					/>
					{errors.lastName && (
						<p>Looks like there was an error: {errors.lastName.type}</p>
					)}
				</div>

				<div>
					<label data-testid='emailLabel' htmlFor='email'>
						Email*
					</label>
					<input
						data-testid='emailInput'
						name='email'
						placeholder='Email'
						ref={register({
							required: true,
							pattern: /^\w+[\w-\.]*\@\w+((-\w+)|(\w*))\.[a-z]{2,3}$/,
						})}
					/>
					{errors.email && (
						<p>Looks like there was an error: {errors.email.type}</p>
					)}
				</div>
				<div>
					<label htmlFor='message' data-testid='messageLabel'>
						Message (Optional)
					</label>
					<textarea
						name='message'
						ref={register({ required: false })}
						data-testid='messageTextarea'
					/>
				</div>

				<label htmlFor='terms' data-testid='checkLabel'>
					I've Read and Agree to Terms And Conditions
				</label>
				<input
					type='checkbox'
					name='terms'
					data-testid='check'
					ref={register({ required: true })}
				/>
				{data && (
					<pre style={{ textAlign: 'left', color: 'white' }} data-testid='pre'>
						{JSON.stringify(data, null, 2)}
					</pre>
				)}
				<input
					type='submit'
					data-testid='submitButton'
					disabled={!formState.isValid}
				/>
			</form>
		</div>
	);
};

export default ContactForm;
