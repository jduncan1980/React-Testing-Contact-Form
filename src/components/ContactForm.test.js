import React from 'react';
import {
	render,
	cleanup,
	fireEvent,
	waitFor,
	act,
} from '@testing-library/react';
import ContactForm from './ContactForm';

afterEach(() => {
	cleanup();
});

test('renders without crashing', () => {
	const { asFragment } = render(<ContactForm />);
	expect(asFragment()).toMatchSnapshot();
});

test('renders first name input and label', () => {
	const { getByTestId } = render(<ContactForm />);

	getByTestId('firstNameLabel');
	getByTestId('firstNameInput');
});

test('renders last name input and label', () => {
	const { getByTestId } = render(<ContactForm />);

	getByTestId('lastNameLabel');
	getByTestId('lastNameInput');
});

test('renders message textarea and label', () => {
	const { getByTestId } = render(<ContactForm />);

	getByTestId('messageLabel');
	getByTestId('messageTextarea');
});

test('renders disabled submit button ', async () => {
	const { getByTestId } = render(<ContactForm />);

	const button = getByTestId('submitButton');
	await waitFor(() => expect(button).toBeDisabled());
});

test('fills out form correctly and submits', async () => {
	const { getByTestId } = render(<ContactForm />);

	const firstName = getByTestId('firstNameInput');
	fireEvent.change(firstName, { target: { value: 'joseph' } });
	expect(getByTestId('firstNameInput')).toHaveValue('joseph');

	const lastName = getByTestId('lastNameInput');
	fireEvent.change(lastName, { target: { value: 'johnson' } });
	expect(getByTestId('lastNameInput')).toHaveValue('johnson');

	const email = getByTestId('emailInput');
	fireEvent.change(email, { target: { value: 'joejohnson@email.com' } });
	expect(getByTestId('emailInput')).toHaveValue('joejohnson@email.com');

	const message = getByTestId('messageTextarea');
	fireEvent.change(message, { target: { value: 'a message' } });
	expect(getByTestId('messageTextarea')).toHaveValue('a message');

	act(async () => {
		const button = getByTestId('submitButton');
		fireEvent.click(button);
	});

	const pre = await waitFor(() => getByTestId('pre'));
	expect(pre).toHaveTextContent('joseph');
	expect(pre).toHaveTextContent('johnson');
	expect(pre).toHaveTextContent('joejohnson@email.com');
	expect(pre).toHaveTextContent('a message');
});

// test('fills out form incorrectly and fails to submit', async () => {
// 	const { getByTestId } = render(<ContactForm />);

// 	const firstName = getByTestId('firstNameInput');
// 	fireEvent.change(firstName, { target: { value: 'joseph' } });
// 	expect(getByTestId('firstNameInput')).toHaveValue('joseph');

// 	const lastName = getByTestId('lastNameInput');
// 	fireEvent.change(lastName, { target: { value: 'johnson' } });
// 	expect(getByTestId('lastNameInput')).toHaveValue('johnson');

// 	const email = getByTestId('emailInput');
// 	fireEvent.change(email, { target: { value: 'joejohnson@email.com' } });
// 	expect(getByTestId('emailInput')).toHaveValue('joejohnson@email.com');

// 	const message = getByTestId('messageTextarea');
// 	fireEvent.change(message, { target: { value: 'a message' } });
// 	expect(getByTestId('messageTextarea')).toHaveValue('a message');

// 	act(() => {
// 		const submit = getByTestId('submitButton');
// 		fireEvent.click(submit);
// 	});

// 	const pre = await waitFor(() => getByTestId('pre'));
// 	expect(pre).toHaveTextContent('joseph');
// 	expect(pre).toHaveTextContent('johnson');
// 	expect(pre).toHaveTextContent('joejohnson@email.com');
// 	expect(pre).toHaveTextContent('a message');
// });
