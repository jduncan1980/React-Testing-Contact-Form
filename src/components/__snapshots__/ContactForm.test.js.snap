// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <form
      data-testid="form"
    >
      <div>
        <label
          data-testid="firstNameLabel"
          for="firstName"
        >
          First Name*
        </label>
        <input
          data-testid="firstNameInput"
          name="firstName"
          placeholder="First Name"
        />
      </div>
      <div>
        <label
          data-testid="lastNameLabel"
          for="lastName"
        >
          Last Name*
        </label>
        <input
          data-testid="lastNameInput"
          name="lastName"
          placeholder="Last Name"
        />
      </div>
      <div>
        <label
          data-testid="emailLabel"
          for="email"
        >
          Email*
        </label>
        <input
          data-testid="emailInput"
          name="email"
          placeholder="Email"
        />
      </div>
      <div>
        <label
          data-testid="messageLabel"
          for="message"
        >
          Message (Optional)
        </label>
        <textarea
          data-testid="messageTextarea"
          name="message"
        />
      </div>
      <input
        data-testid="submitButton"
        type="submit"
      />
    </form>
  </div>
</DocumentFragment>
`;
